compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -626,18 +626,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     }
                 };
 
-                let coerce_to = match opt_coerce_to {
-                    Some(c) => c,
-                    None => {
-                        // If the loop context is not a `loop { }`, then break with
-                        // a value is illegal, and `opt_coerce_to` will be `None`.
-                        // Return error in that case (#114529).
-                        return Ty::new_misc_error(tcx);
-                    }
-                };
+                // If the loop context is not a `loop { }`, then break with
+                // a value is illegal, and `opt_coerce_to` will be `None`.
+                // Set expectation to error in that case.
+                let coerce_to = opt_coerce_to.unwrap_or_else(|| Ty::new_misc_error(tcx));
 
                 // Recurse without `enclosing_breakables` borrowed.
                 e_ty = self.check_expr_with_hint(e, coerce_to);
+
+                // If break with value was found illegal above,
+                // return error right here or we risk passing typeck
```

Comment:
It doesn't matter if we pass typeck. We've already delayed a bug. What happens if you remove this?

Reply:
Alternatively, you should taint the infcx in that `unwrap_or_else` above.

Reply:
> It doesn't matter if we pass typeck. We've already delayed a bug. What happens if you remove this?

If we pass typeck here it results in an ICE during MIR interpretation as reported in #114529. I had created PR #117382 to fix #114529. However, that resulted in regression #117821. The present PR intends to fix that regression.

Reply:
> Alternatively, you should taint the infcx in that `unwrap_or_else` above.

You're right, tainting infcx in `unwrap_or_else` is much cleaner. I've done that now.